qiskit_ode/signals/signals.py

```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
+
+        # set carrier and phase
+        self.carrier_freq = carrier_freqs
+        self.phase = phases
+
+    def envelope(self, t: Union[float, np.array, Array]) -> Array:
+        """Evaluate envelopes of each component. For vectorized operation,
+        last axis indexes the envelope, and all proceeding axes are the
+        same as the ``t`` arg.
+        """
+        # to do: jax version
+        # not sure what the right way to do this is, here we actually need
+        # to get it to use np/jnp
+        return np.moveaxis(Array([env(t) for env in self._envelopes]), 0, -1)
+
+    def complex_value(self, t: Union[float, np.array, Array]) -> Array:
+        exp_phases = np.exp(np.expand_dims(Array(t), -1) * self._carrier_arg + self._phase_arg)
+        return np.sum(self.envelope(t) * exp_phases, axis=-1)
+
+    def __len__(self):
+        return len(self.components)
+
+    def __getitem__(self, idx: Union[int, List, np.array, slice]) -> Signal:
+        """Enables numpy-style subscripting, as if this class were a 1d array."""
+
+        if type(idx) == np.ndarray and idx.ndim > 0:
+            idx = list(idx)
+
+        sublist = None
+        # get a list of the subcomponents
+        if type(idx) == list:
+            # handle lists
+            sublist = operator.itemgetter(*idx)(self.components)
+
+            # output will either be a single signal or a tuple of Signals
+            # convert to list if tuple
+            if type(sublist) == tuple:
+                sublist = list(sublist)
+        else:
+            # handle slices or singletons
+            sublist = operator.itemgetter(idx)(self.components)
 
-    # ensure both arguments are signals
-    if isinstance(sig1, (int, float, complex)):
-        sig1 = Constant(sig1)
+        # sublist should either be a single Signal, or a list of Signals
+        if isinstance(sublist, Signal):
+            return sublist
+        else:
+            return SignalSum(*sublist)
 
-    if isinstance(sig2, (int, float, complex)):
-        sig2 = Constant(sig2)
+    def __str__(self):
+        if self.name is not None:
+            return str(self.name)
 
-    return Signal(
-        lambda t: sig1.envelope_value(t) * sig2.envelope_value(t),
-        sig1.carrier_freq + sig2.carrier_freq,
-        sig1.phase + sig2.phase,
-    )
+        if len(self) == 0:
+            return "SignalSum()"
 
+        default_str = str(self[0])
+        for sig in self.components[1:]:
+            default_str += " + {}".format(str(sig))
 
-def signal_add(
-    sig1: Union[BaseSignal, float, int, complex], sig2: Union[BaseSignal, float, int, complex]
-) -> BaseSignal:
-    r"""Implements mathematical addition between two signals.
-    Since a signal is represented by
-    .. math::
+        return default_str
 
-        \Omega(t)*exp(2\pi i \nu t)
+    def flatten(self) -> Signal:
+        """Merge into a single ``Signal``. The output frequency is given by the
+        average.
+        """
```

Comment:
Would it make sense to have the average frequency as the default but then let the user override this? For example, with a mixer I probably do not want to have the average as the carrier frequency. We could also think of methods to have the carrier at the frequency where the power spectral density of the signal is highest.

Reply:
I agree, though I think this requires enough thought that it warrants its own subsequent PR. Currently I'm imagining:
- `flatten` is a brute force function that forces a sum into a single `Signal`, and should be avoided where possible. I view this mainly as a stopgap for the near-term until we update the rest of the repo to work with `SignalSum` objects.
- I think we should add a `simplify` method (in a subsequent PR) that more cleverly attempts to merge signals in a sum based on either their PSD, or on their `carrier_freq` values (and we can have a tolerance here).

I think for AWG implementation we want to use `simplify`, so that it doesn't merge components that are too far apart in frequency space.